Warning for composite experiment run options

coruscating opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Addresses #96 with simple approach, just giving a blanket warning when sub-experiments have non-trivial run options.


### Details and comments
Should the dummy backend and experiment class be its separate file? Do we ever need it for other tests?

yaelbh reviewed (changes requested):
Thanks for addressing the issue. A few small comments:
1. Can you please do it also with transpile options.
2. Regarding your question about the dummy experiment: you can consult with @eggerdj, since the data processor tests have a fake experiment, which looks similar.
3. Maybe test also that there is no over-triggering.

coruscating commented:
@yaelbh Thanks for your comments. After talking to @eggerdj, I've moved `fake_experiment.py` from `data_processing` up a level for all tests to use, and added `fake_backend.py`. I also added tests for all four options.

I manually tested that warnings only print once even if it's applicable for multiple sub-experiments, that's the default behavior of python warnings, but I'm not sure how to test that with `assertWarns`.

yaelbh commented on qiskit_experiments/composite/composite_experiment.py:
```diff
@@ -63,6 +66,10 @@ def _add_job_metadata(self, experiment_data, job, **run_options):
             sub_exp = self.component_experiment(i)
+            if sub_exp.run_options != Options() or sub_exp.transpile_options != Options():
+                warnings.warn(
+                    "All sub-experiment options are overridden by composite experiment settings."
```
Analysis options are not overridden... :-)

coruscating replied:
Good point, I've updated the warning :)

yaelbh reviewed (approved):
I left only one small comment

yaelbh commented on qiskit_experiments/composite/composite_experiment.py:
```diff
@@ -63,6 +66,15 @@ def _add_job_metadata(self, experiment_data, job, **run_options):
+            if (
+                sub_exp.run_options != Options()
+                or sub_exp.transpile_options != Options()
+                or sub_exp.experiment_options != Options()
```
Experiment options are not necessarily overridden. I think this one depends on the specific experiment and the specific option. For example for T1 you can specify a different set of delays to different qubits and the experiment works well on a device.

coruscating replied:
Yes, I noticed that the T1 tests trigger the warning but work fine. I'll see if I can change the condition here to only warn if the options are actually overridden.

coruscating replied:
I updated the code to check the transpile and run options against the default ones now, and got rid of the experiment options warning. Because each experiment is different like you said, I couldn't see an easy way to check for sure whether the experiment options are overridden. But for the vast majority of experiments, experiment options should be doing something to the circuits that are not overridden, so I decided to not warn.

eggerdj commented on test/fake_experiment.py:
```diff
@@ -0,0 +1,46 @@
+        return [], None
+
+
+class FakeExperiment(BaseExperiment):
+    """Fake experiment class for testing."""
+
+    __analysis_class__ = FakeAnalysis
```
I'm okay with moving this here. Note that there is a similar class in `test_curve_fit.py` perhaps worth consolidating all these testing classes in one location to avoid duplication.

coruscating replied:
Good catch, thanks, I've combined the class.

yaelbh commented on qiskit_experiments/composite/composite_experiment.py:
```diff
@@ -63,6 +66,23 @@ def _add_job_metadata(self, experiment_data, job, **run_options):
             sub_exp = self.component_experiment(i)
+
+            # Construct default transpile options with initial layout
+            default_transpile_options = Options(**sub_exp._default_transpile_options().__dict__)
```
Shorter:
```
default_transpile_options = sub_exp._default_transpile_options()
```

yaelbh commented on qiskit_experiments/composite/composite_experiment.py:
```diff
@@ -63,6 +66,23 @@ def _add_job_metadata(self, experiment_data, job, **run_options):
+            default_transpile_options = Options(**sub_exp._default_transpile_options().__dict__)
+            default_transpile_options.update_options(
+                **{"initial_layout": sub_exp.transpile_options.get("initial_layout")}
+            )
```
These lines reflect something that's maybe not done correctly in `BaseExperiment`. In `BaseExperiment`, the default transpile options don't include the initial layout, which is added to the non-default transpile options in `BaseExperiment.__init__`. Perhaps the initial layout could be part of the default transpile options.

coruscating replied:
Agreed, that makes a lot more sense.

chriseclectic replied:
I agree with @yaelbh that this looks like an issue with how layout is handled in base experiment. One solution would be to remove initial layout from the `BaseExperiment.__init__` entirely and only add it to the transpile options just before calling transpile in `run`:

```python
# Generate and transpile circuits
transpile_opts = self.transpile_options.__dict__
transpile_opts['initial_layout'] = list(self._physical_qubits)
circuits = transpile(self.circuits(backend), backend, **transpile_opts)
```
Then no changes should be needed for composite experiment with regards to the layout (and it would be optional to leave the current exception if you try and set initial_layout in transpile options since it is always going to be overriden anyway).

coruscating commented:
@yaelbh I've updated the PR by moving `initial_layout` as @chriseclectic suggested.
